### PR TITLE
Fix #268/#71: HMR support for web sockets

### LINKED ISSUE #71: Handle websocket in dev-mode or document that it is not forwarded (HMR)
Some NodeJS dev-server don't allow to set the websocket port for for hot reload. Currently those websocket connection are not forwarded by Quinoa.

Some other dev-server like Vite allow to change the HMR port, then http is sent and forwarded by Quinoa and HMR is sent to the frontend dev-server (which works great).

### LINKED ISSUE #268: Add support for React NextJS
Reproducer project: 
[quinoa-nextjs.zip](https://github.com/quarkiverse/quarkus-quinoa/files/10936515/quinoa-nextjs.zip)

NextJS serves on port 3000 like Create React app does and it does not serve an index.html but I have Quinoa configured where it almost works.

```properties
quarkus.quinoa.package-manager-install=true
quarkus.quinoa.package-manager-install.node-version=16.17.0
quarkus.quinoa.dev-server.port=3000
quarkus.quinoa.enable-spa-routing=true
quarkus.quinoa.build-dir=out
quarkus.quinoa.index-page=/
```

If you run `mcn quarkus:dev` on the above project the one remaining issue is that NextJS opens a websocket this works normally and returns a 101.
image

But in Quinoa its return a 403 and tries over and over again.  I believe Next JS needs this socket open to communicate changes.

### PR DESCRIPTION
**Pull request by melloware**
Fix #268 Netx JS support
Fix #71 Handle websockets in Dev Mode

### CODE REVIEW

**Comment by melloware** [comment]
Tested with both Vite and NextJS successfully.

**Comment by melloware** [comment]
Any websocket you want ignored you can still use the ignored oaths peppery to skip!

File: runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java [hunk 1]
```diff
@@ -58,8 +67,73 @@ public void handle(final RoutingContext ctx) {
             next(currentClassLoader, ctx);
             return;
         }
-        final String uri = computeURI(resourcePath, request);
+
         final MultiMap headers = request.headers();
+        final String uri = computeURI(resourcePath, request);
+
+        // WebSocket connection for Hot Module Reloading (HMR)
+        if (!config.prodMode && headers.get("Connection").contains("Upgrade")) {
```
Comment (ia3andy) [inline runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java:75]: headers.get("Connection") can be null right?

Comment (ia3andy) [inline runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java:75]: could we delegate to handling methods for classic request and websocket to make this method a bit shorter

File: runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java [hunk 2]
```diff
@@ -58,8 +67,73 @@ public void handle(final RoutingContext ctx) {
...
+                        .setURI(forwardUri)
+                        .setHeaders(socketHeaders)
+                        .setSubProtocols(subProtocols)
+                        .setAllowOriginHeader(false);
+
+                LOG.infof("Quinoa client websocket: %s:%s%s", host, port, forwardUri);
+                httpClient.webSocket(options, clientContext -> {
+                    if (clientContext.succeeded()) {
+                        try {
+                            final WebSocket clientWs = clientContext.result();
```
Comment (ia3andy) [inline runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java:105]: What if it's a websocket connection made to be handled by the Quarkus backend? (and not a HMR)

File: runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java [hunk 3]
```diff
@@ -58,8 +67,73 @@ public void handle(final RoutingContext ctx) {
...
         }
-        final String uri = computeURI(resourcePath, request);
+
         final MultiMap headers = request.headers();
+        final String uri = computeURI(resourcePath, request);
+
+        // WebSocket connection for Hot Module Reloading (HMR)
+        if (!config.prodMode && headers.get("Connection").contains("Upgrade")) {
+            // Vite uses the root URL "/" which is special
+            final String forwardUri = request.uri().equals("/") ? "/" : uri;
```
Comment (ia3andy) [inline runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java:77]: Are you sure this has no side-effect in other cases?

**Review by ia3andy — CHANGES_REQUESTED** [review]
Thanks, that's very cool, I've just added a few comments.. I will give it a try later 👍

File: runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java (hunk 1 above)
Comment (melloware) [reply runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java]: I refactored to delegate to two different methods.

File: runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java (hunk 2 above)
Comment (melloware) [reply runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java]: You can still use your `ignorePaths` value to ignore any WebSocket connection you don't want touched.   But I have a feeling this code is going to work no matter what.

File: runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java (hunk 3 above)
Comment (melloware) [reply runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java]: It shouldn't its just this special case where Vite uses "/" but if I use your converted Uri its "/index.html" which then fowards to the wrong URL for the web socket.

**Comment by melloware** [comment]
Very happy with this version now, more graceful and removes Vite hack

**Comment by ia3andy** [comment]
I've test vite+svelte and get this error:
```
2023-03-14 10:13:37,160 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2023-03-14 10:13:37,161 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, quinoa, resteasy-reactive, smallrye-context-propagation, vertx]
2023-03-14 10:13:42,405 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Sec-WebSocket-Protocol: vite-hmr
2023-03-14 10:13:42,405 ERROR [io.ver.cor.net.imp.ConnectionBase] (vert.x-eventloop-thread-2) Invalid object PooledUnsafeDirectByteBuf(freed)
2023-03-14 10:13:42,406 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa is forwarding web socket: '/'
2023-03-14 10:13:42,433 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa client websocket: 127.0.0.1:5173/
2023-03-14 10:13:42,451 WARN  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Server websocket is not ready: WebSocket is closed
2023-03-14 10:13:43,058 ERROR [io.ver.cor.net.imp.ConnectionBase] (vert.x-eventloop-thread-2) Invalid object PooledUnsafeDirectByteBuf(freed)
2023-03-14 10:13:43,060 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Sec-WebSocket-Protocol: vite-hmr
2023-03-14 10:13:43,061 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding web socket: '/'
2023-03-14 10:13:43,063 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa client websocket: 127.0.0.1:5173/
2023-03-14 10:13:43,071 WARN  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Server websocket is not ready: WebSocket is closed
```

Here is the project: https://github.com/ia3andy/vite-svelte-quinoa

**Comment by melloware** [comment]
Thanks for the reproducer I will investigate and let you know!

**Comment by melloware** [comment]
Weird your reproducer is working fine for me and HMR is working great!

```
Microsoft Windows [Version 10.0.22621.1265]

Apache Maven 3.9.0 (9b58d2bad23a66be161c4664ef21ce219c2c8584)
Maven home: C:\Tools\apache-maven-3.9.0
Java version: 17.0.5, vendor: Eclipse Adoptium, runtime: C:\Tools\jdk-17
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 11", version: "10.0", arch: "amd64", family: "windows"
```

**Comment by melloware** [comment]
Its also strange yours called the websocket twice and mine only called it once correctly.

```
2023-03-14 07:30:16,886 INFO  [io.quarkus] (Quarkus Main Thread) vite-svelte-quinoa 1.0.0-SNAPSHOT on JVM (powered by Quarkus 2.16.4.Final) started in 16.237s. Listening on: http://localhost:8080

2023-03-14 07:30:16,887 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2023-03-14 07:30:16,888 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, quinoa, resteasy-reactive, smallrye-context-propagation, vertx]
2023-03-14 07:30:30,103 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Sec-WebSocket-Protocol: vite-hmr
2023-03-14 07:30:30,104 ERROR [io.ver.cor.net.imp.ConnectionBase] (vert.x-eventloop-thread-2) Invalid object PooledUnsafeDirectByteBuf(freed)
2023-03-14 07:30:30,105 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa is forwarding web socket: '/'
2023-03-14 07:30:30,132 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa client websocket: 127.0.0.1:5173/
2023-03-14 07:30:30,150 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa has forwarded web socket: '/'
```

**Comment by melloware** [comment]
Also from looking at your threads it looks like the socket got called from thread 2 and 3 simultaneously from your browser?

**Comment by ia3andy** [comment]
I am using chrome

**Comment by ia3andy** [comment]
ah maybe I had two tabs open

**Comment by melloware** [comment]
Yep try it fresh from icognito mode just to take all caching out of the equation and starts a fresh websocket.

**Comment by ia3andy** [comment]
```

2023-03-14 14:56:31,706 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2023-03-14 14:56:31,707 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, quinoa, resteasy-reactive, smallrye-context-propagation, vertx]
2023-03-14 14:56:37,448 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Sec-WebSocket-Protocol: vite-hmr
2023-03-14 14:56:37,449 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding web socket: '/'
2023-03-14 14:56:37,467 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa client websocket: 127.0.0.1:5173/
2023-03-14 14:56:37,470 ERROR [io.ver.cor.net.imp.ConnectionBase] (vert.x-eventloop-thread-2) Invalid object PooledUnsafeDirectByteBuf(freed)
2023-03-14 14:56:37,471 ERROR [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa failed to forward request '/node_modules/svelte-hmr/runtime/overlay.js?v=446aea97', see logs.: java.lang.IllegalStateException: Invalid object PooledUnsafeDirectByteBuf(freed)
	at io.vertx.core.http.impl.Http1xClientConnection.lambda$static$0(Http1xClientConnection.java:79)
	at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:721)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:157)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)


14:56:37 [vite] ws error:
RangeError: Invalid WebSocket frame: RSV1 must be clear
    at Receiver.getInfo (file:///Users/ia3andy/Downloads/vite-svelte-quinoa/src/main/webui/node_modules/vite/dist/node/chunks/dep-ca21228b.js:54761:14)
    at Receiver.startLoop (file:///Users/ia3andy/Downloads/vite-svelte-quinoa/src/main/webui/node_modules/vite/dist/node/chunks/dep-ca21228b.js:54708:22)
    at Receiver._write (file:///Users/ia3andy/Downloads/vite-svelte-quinoa/src/main/webui/node_modules/vite/dist/node/chunks/dep-ca21228b.js:54646:10)
    at writeOrBuffer (node:internal/streams/writable:391:12)
    at _write (node:internal/streams/writable:332:10)
    at Receiver.Writable.write (node:internal/streams/writable:336:10)
    at Socket.socketOnData (file:///Users/ia3andy/Downloads/vite-svelte-quinoa/src/main/webui/node_modules/vite/dist/node/chunks/dep-ca21228b.js:57428:37)
    at Socket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
2023-03-14 14:56:37,513 WARN  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Server websocket is not ready: WebSocket is closed
2023-03-14 14:56:40,732 ERROR [io.ver.cor.net.imp.ConnectionBase] (vert.x-eventloop-thread-3) Invalid object PooledUnsafeDirectByteBuf(freed)
2023-03-14 14:56:40,734 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Sec-WebSocket-Protocol: vite-hmr
2023-03-14 14:56:40,735 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa is forwarding web socket: '/'
2023-03-14 14:56:40,736 ERROR [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa failed to forward request '/node_modules/.vite/deps/chunk-XTA3UROE.js?v=446aea97', see logs.: io.vertx.core.http.HttpClosedException: Connection was closed

2023-03-14 14:56:40,738 INFO  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa client websocket: 127.0.0.1:5173/
2023-03-14 14:56:40,743 WARN  [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Server websocket is not ready: WebSocket is closed

**Comment by melloware** [comment]
So weird that is not happening for me with your project and Svelte is hot reloading fine. I can make a change in a page and see it reflected immediately and see the traffic on the websocket.

**Comment by ia3andy** [comment]
Any hint on how I could debug it to figure it out?

**Comment by melloware** [comment]
So I have DEBUG messages in all the socket send and receive you can flip them to INFO to see any WebSocket traffic.

**Comment by ia3andy** [comment]
> Interesting your error `RSV1 must be clear` is related to compressing the socket: [websockets/ws#1140](https://github.com/websockets/ws/issues/1140)
> 
> Are you using a MacOS Chrome by any chance?

yes!

**Comment by ia3andy** [comment]
It's working on Safari but the problem is in the Quinoa WS code as it works with Chrome if I add this:
```
server: {
    port: 5173,
    host: '127.0.0.1',
    hmr: {
      port: 5173,
      host: '127.0.0.1',
    }
  }
  ```

**Comment by melloware** [comment]
OK what Node version are you using?

**Comment by ia3andy** [comment]
With the reproducer project, it's the one specified in the application.properties `16.17.0`

**Comment by melloware** [comment]
oK yeah I tried 16.17 and 18.8.0 and I can't reproduce your issue .

**Comment by melloware** [comment]
OK I pushed a fix not to foward headers on to Vite.  Can you try again?

**Comment by ia3andy** [comment]
It's still failing but I have been debugging all morning and I made a bit of progress.. I'll let you know..

**Comment by melloware** [comment]
Awesome.  I appreciate it. I wish I could make it happen regularly so you didn't have to debug it!

**Comment by melloware** [comment]
testing it now!

**Review by ia3andy — APPROVED** [review]

**Comment by melloware** [comment]
OK tested NextJS and Vite on Windows and both looking good.  The only strange thing I am seeing is a bunch of left TIMED_WAIT on socket port 3000 I wasn't noticing before.

image

Even after I kill the app those socket ports are still hanging with PID 0 listed.

**Comment by melloware** [comment]
I think this is safe to merge!

**Comment by melloware** [comment]
Commits Squashed

**Comment by ia3andy** [comment]
Thanks @melloware this is a very good contribution!!!

**Comment by ia3andy** [comment]
@all-contributors please add @melloware for code, doc

**Pull request merged**